libsupport/include/katana/ProgressTracer.h

```diff
@@ -0,0 +1,225 @@
+//   and scope3's ProgressSpan finishing before scope1's if there is no error
+
+class KATANA_EXPORT ProgressTracer {
+  friend class ProgressSpan;
+
+  void SetSpan(std::shared_ptr<ProgressSpan> span);
+
+protected:
+  ProgressTracer(uint32_t host_id, uint32_t num_hosts)
+      : host_id_(host_id), num_hosts_(num_hosts) {}
+
+public:
+  virtual ~ProgressTracer() = default;
+
+  static std::shared_ptr<ProgressTracer> GetProgressTracer() { return tracer_; }
+  static void SetProgressTracer(std::shared_ptr<ProgressTracer> tracer);
+
+  /// Create a new top level span if one does not already exist.
+  /// Otherwise creates a child span of the current span
+  /// Can specify to not finish the current scope’s span by RAII by
+  /// setting finish_on_close=false
+  ProgressScope StartActiveSpan(
+      const std::string& span_name, bool finish_on_close = true);
+  ProgressScope StartActiveSpan(
+      const std::string& span_name,
+      const std::unique_ptr<ProgressContext>& child_of,
+      bool finish_on_close = true);
+
+  /// Set the current active thread-local span to provided span
+  /// Returns a ProgressScope
+  /// Can specify to not finish the span when the scope goes out of scope
+  /// by RAII by setting finish_on_close=false
+  /// If called with a nullptr span,
+  ///    calls StartActiveSpan("error: nullptr span given to SetActiveSpan", finish_on_close)
+  ProgressScope SetActiveSpan(
+      std::shared_ptr<ProgressSpan> span, bool finish_on_close = true);
+
+  /// Create a new top level span if ignore_active_span=true and
+  /// no child_of value is given
+  /// Otherwise creates a child span of the child_of span or active span
+  /// Should only be used to handle multiple active spans simultaneously
+  virtual std::shared_ptr<ProgressSpan> StartSpan(
+      const std::string& span_name, bool ignore_active_span = false) = 0;
+  virtual std::shared_ptr<ProgressSpan> StartSpan(
+      const std::string& span_name, std::shared_ptr<ProgressSpan> child_of) = 0;
+  virtual std::shared_ptr<ProgressSpan> StartSpan(
+      const std::string& span_name,
+      const std::unique_ptr<ProgressContext>& child_of) = 0;
+
+  // For passing spans across process/host boundaries
+  // These functions are needed, but are implemented now for
+  // debugging purposes, they will be replaced
+  // Extract returns a nullptr on failure
+  virtual std::string Inject(const ProgressContext& ctx) = 0;
+  virtual std::unique_ptr<ProgressContext> Extract(
+      const std::string& carrier) = 0;
+
+  /// Get the current scope’s span to add tagging/logging without having
+  /// to pass around scopes/spans as parameters
+  virtual std::shared_ptr<ProgressSpan> GetActiveSpan();
+  uint32_t GetHostId() { return host_id_; }
+  uint32_t GetNumHosts() { return num_hosts_; }
+
+  /// Close is called when a tracer is finished processing spans
+  /// This should be called to ensure any and all buffered spans are
+  /// flushed
```

Comment:
It's dependent on subclass implementations but it is required by Jaeger and Lightstep which I've been looking at so far, probably all mature tracers require it to behave correctly

I don't think it would be an issue to have this called in the destructor, but the tracer of a program will be stored in a static variable.  Not too sure about C++ destructor semantics w.r.t. this but if that's fine then I can just move it to the destructor

Reply:
There are two interacting issues.

1. How the object cleans itself up.
```cpp
// RAII
{
  Foo foo;
}

// Explicit management
{
  Foo foo;
  foo.Close();
}

// Hybrid (Close when out of scope and not already closed)
{
  Foo foo;
  //foo.Close();
}
```

2. How singletons are managed.

```cpp
// 1. Static variable 
Foo foo;

// 2. Static class variable
Foo Foo::foo;

// 3. Managed pointer
std::unique_ptr<Foo> foo;
```

The issue with both 1 and 2 is that when exactly `foo` is constructed and destructed is indeterminate in C++.  The issue with 3 is that you need to know when to construct (and probably destruct) foo, i.e., the construction must precede any uses and the destruction must be after the last use. This becomes difficult if there are references to foo in other static variable constructors.

I'd prefer 3 if possible because once we have to deal ordering of static initializers, it is hard to undo that.

Reply:
I messed something up with Github and had to post this below as well:

Right now we are using a static class variable shared_ptr to store the global tracer, eventually this will be thread local
Then we construct it in DistMemSys or SharedMemSys
And currently we have no destructor logic

Reply:
See https://github.com/KatanaGraph/katana/pull/370#discussion_r678598236 for some continued discussion of singletons.